### ISSUE TITLE
Add GraphQL guides + examples to the site

### ISSUE THREAD

**Issue by samselikoff** [miragejs/site#264]
Now that v0.1.0 of https://github.com/miragejs/graphql is out, we should integrate it into the site so folks can find about it more easily!

TODOS:

- [x] Add a guide to the site
- [x] [REPL] Add support for `graphql` library (get [Sandbox 230](https://miragejs.com/repl/v1/230) working)
- [x] [REPL] Add support for template string (get [Sandbox 231](https://miragejs.com/repl/v1/231) working)
- [ ] ~[REPL] Add support for string request body (get [Sandbox 232](https://miragejs.com/repl/v1/232) working)~

## Add a guide to the site

To start, I think we should create a single new guide page on the site under Advanced, called GraphQL. And it should have basically a quickstart/overview, that shows off how @miragejs/graphql works. Then, a link to the repo.

I think this will let Rocky/us/anyone else keep iterating on @miragejs/graphql while it's new, but give folks a way to find out about it on the site.

## Add GraphQL support to the REPL

The second main piece is getting GraphQL support in https://miragejs.com/repl. There's a lot we can do here but I'd like us to focus on the smallest steps possible in order, to get us some forward momentum.

The first is that the `graphql` library currently doesn't work in the REPL. You can see from [Sandbox 230](miragejs.com/repl/v1/230) that there's a runtime error, `process is not defined`. It seems `graphql` is trying to access the node process (which seems strange to me because I've used this in client Mirage code before). Anyway, this is as far as Ryan and I got last time we were investigating this, so this is really the first thing that needs to be explored before GraphQL can work at all in the REPL.

Once 230 works, the next improvement would be to add template string support, so the schema can be written using multiline backticks. That's [Sandbox 231](miragejs.com/repl/v1/231). The reason template strings don't work right now is that the Server config input currently does not have full ES6 support. The library we're using to evaluate the user's input is called plnker, and we haven't dug deep into how it's configured. But it seems certain ES6 features are supported while others are not. Unfortunately, ES6 template strings (back ticks) are not supported, which makes any GraphQL example unpalatable. So, I think the next thing should be figuring out how to get template strings supported in Plunkr. This is an unknown unknown, I don't know what it involves or even if it's possible. Hopefully there's a way to do it, but there's a possibility we may have to end up moving to our own bundling solution with something like Rollup. Plunkr has allowed us to kick this can down the road for now.

Last would be getting [Sandbox 232](miragejs.com/repl/v1/232) working, which uses a GraphQL Query as the request body, instead of a JSON object with a query key and a GraphQL query as the value. This should be the easiest fix, because we just have some of our own validation on the request body input field at the moment ensuring it's JSON, but that's just there to help folks who are using Mirage to handle normal JSON payloads. So, I think we should remove that validation so raw text can be used to make requests.

There's of course a ton more we can do to make the GraphQL REPL experience better but I think this is where we should start!

**Comment by jneurock** [comment]
Hey @samselikoff, quick question on the client half of the REPL...

Would it make sense to have tabs at the top level, one for REST and the other for GraphQL? To use it as-is, I would need to select POST, fill in the URL and then put whatever GraphQL expects in the request body field.

As a user, I would rather have a GraphQL tailored client experience where I can just enter the URL and then put a query into a multiline text field. I'd be happy to work on this but I noticed there's already some hidden work in that area and it would be good to sync up so I don't trample on any WIP. WDYT?

Alternatively, we could try putting [GraphiQL](https://github.com/graphql/graphiql/tree/main/packages/graphiql#readme) into that space for a GraphQL experience but it has a lot of features and I have major doubts it would be a good experience.

**Comment by samselikoff** [comment]
I absolutely want to get there, but I'd prefer to keep the scope smaller for now and layer those on later. I have some other changes I want to make to the overall layout too so it would be worth taking it all into consideration before jumping into any implementation 👍

tl;dr I want to keep improving, but if the current UI supported GraphQL today that would be much better than not having it for a month or two (as the larger scoped work can stall or drag out).

**Comment by jneurock** [comment]
Gotcha. I was definitely trying to think in terms of small scope but with your input here I think I can attempt to make some subtle changes to get GraphQL support in there without worrying about layouts or anything large.

**Comment by samselikoff** [comment]
Cool, yeah I tried to lay it out in the 3 TODOs above. I think the work should be done in that order, the first one being getting Sandbox 230 which is just figuring out why `graphql` isn't working on the client. I think the only visual UI changes we need is for the 3rd todo, to allow for a raw string request body.

**Comment by jneurock** [comment]
Related to Sandbox 230... See https://github.com/graphql/graphql-js/issues/2676. Looks like there's a plan to fix this in the next *major* release of GraphQL JS. In the meantime, I have an idea of how to work around this.

**Comment by samselikoff** [comment]
Haha wow – nice sleuthing!

Just verified it works: https://deploy-preview-813--miragejs.netlify.app/repl/v1/239

Super exciting!

**Comment by samselikoff** [comment]
@jneurock Just loaded the example I made with your latest PR and it looks like the back ticks are escaped when they come back from the backend:

https://miragejs.com/repl/v1/245

Maybe we can escape them not directly in the React state but when syncing with the iframe?

**Comment by samselikoff** [comment]
Also another thing, looks like we should get @miragejs/graphql set up with some build tooling because right now the REPL can't work with it:

https://miragejs.com/repl/v1/246

And I think that's because right now it's all unbundled. Maybe we should use TSDX - because getting `miragejs` set up with rollup was kinda a nightmare :/

**Comment by samselikoff** [comment]
(Btw killer work!! Just dropping comments as I find them but I'm super excited by this + thank you for your continued contribs ❤️)

**Comment by jneurock** [comment]
> @jneurock Just loaded the example I made with your latest PR and it looks like the back ticks are escaped when they come back from the backend:
> 
> https://miragejs.com/repl/v1/245
> 
> Maybe we can escape them not directly in the React state but when syncing with the iframe?

Yes, makes total sense. I was looking for that bit of code but couldn't find it. Maybe you can point me to the place where it syncs with the iframe? That would be the best place to escape the backticks.

**Comment by jneurock** [comment]
> Also another thing, looks like we should get @miragejs/graphql set up with some build tooling because right now the REPL can't work with it:
> 
> https://miragejs.com/repl/v1/246
> 
> And I think that's because right now it's all unbundled. Maybe we should use TSDX - because getting `miragejs` set up with rollup was kinda a nightmare :/

I was thinking we should add a build at some point. Do you think you could open an issue in the GraphQL repo and put in some more details? I would have defaulted to copying the setup from Mirage JS but if was kind of a nightmare I can go another route. I don't know what TSDX is but I'm fairly familiar with Webpack, if that's an option.

**Comment by jneurock** [comment]
@samselikoff, RE: sandbox 232, I'm not sure it makes sense for a GraphQL query to be the entire request body. It would save some typing, sure, but it wouldn't reflect what most GraphQL clients would send to the server. The server should typically expect JSON, see https://graphql.org/learn/serving-over-http/#post-request.

There's a note in there about allowing the query to be the entire request body but then the content type of the request needs to be `application/graphql`. We could definitely make things easier by allowing the input to consist of a raw GraphQL query but not without having some sort of GraphQL mode for the REPL so it knows how to make request decisions.

**Comment by samselikoff** [comment]
Sorry for the delay here!

Just to get a refresher:

 - Using JSON to send the query in the `query` key of an object seems ok for now, won't be that surprising to GraphQL users.
 - Regardless, if we want to improve the DX of GraphQL we should invest in a different "mode" for the interface similar to graphiql. But that's a bigger project and we should tackle that after we get `@miragejs/graphql` working.
 - To get `@miragejs/graphql` working we need to figure out the build stuff

Does that all sound correct?

**Comment by jneurock** [comment]
Hey, thanks! Yep, all sounds correct except maybe the last point. I'm not sure `@miragejs/graphql` needs a build. I need to look a little closer into one of its dependencies. From what I can tell, `plnkr` should work fine with it but the `b2a` dependency won't load correctly. It seems to have something to do with [jspm.io](https://jspm.io).

There are 2 ways forward I can imagine:
1. Remove the `b2a` dependency, if I can confirm this is the only problem.
1. Add a build to `@miragejs/graphql`. We may want to do this anyway for some use case I can't think of right now.

**Comment by samselikoff** [comment]
Ok hm. Why don't we start by adding a `module` field to `@miragejs/graphql`'s package.json? Plnkr might respect that when pulling down modules. (We should have that in there anyway because I think a lot of build tools look for that first.)

This is miragejs's: https://github.com/miragejs/miragejs/blob/master/package.json#L6

I believe for `@miragejs/graphql` it could just point to the same `index.js` file that `main` currently does.

**Comment by jneurock** [comment]
I think that's a great guess but I don't think that will help for this particular issue loading the `b2a` dependency. Here's a relevant thread from the jspm.io Discord server. Feel free to ignore this, I just thought it was interesting.

> 07/29/2020
> Hi, i am trying to understand the exports field now. Please bear with the doubt 😓 
> So, if exports field is not available in package.json does *jspm* fallback to files field and serve the files.
> Since, antd design system (https://github.com/ant-design/ant-design/blob/master/package.json) does not specify any exports filed but  files instead.
> I was able to load whole antd using jspm.dev/antd@4.5.0 and at the same time single component alone using jspm.dev/npm:antd@4.5.0/es/button
> But when comes to react-bootstrap it doesn't work that way, as they specify files but the esm bundle is actually present under lib
> So, jspm.dev/react-bootstrap@1.3.0/lib/esm/button doesn't work. By adding 
```js
exports: {
  lib: 'lib',
  esm: 'lib/esm'
}
```
> will allow jspm to work right ? Using jspm.dev/react-bootstrap/esm/button` after adding it. Would like to make PR to their > repo if my assumption is true. Sorry, for the long question  

---

> 07/29/2020
> the "main" is used when "exports" is not provided
> but in addition to the "main" we do depenency analysis
> so basically when there is no "exports", we go through the list of all packages on npm that require or import this package, and we then do a statistical analysis of these to detect the most common patterns
> those are then treated as if they were defined as "exports"
> for react-bootstrap, it looks like the analysis is working fine
> but it is detecting react-bootstrap/button not react-bootstrap/lib/esm/button
> see eg https://unpkg.com/browse/react-bootstrap@1.3.0/
> UNPKG - react-bootstrap
> The CDN for react-bootstrap
> what is published to npm is not what is on GitHub!
> the correct exports for that would be
> `"exports": { ".": "./esm/index.js", "./": "./esm/" }` I believe
> which would then permit `import('react-bootstrap')`  as well as `import('react-bootstrap/Tooltip.js')` etc
> note the .js extensions are needed for path exports

---

>  12:45 PM
> Thanks for the explanation, but i see most of the packages use module to expose their esm modules. So, are there any specific use cases why jspm checks only for main if exports is not defined. I mean if the package is exposing module then it can be used as entry for esm bundles 
> Ah, got the answer from you Github comment. Thank you   "Also, jspm does not use the "module" package.json entry because this is also not supported in Node.js."

---

> Yesterday at 4:21 PM
> "module" doesn't follow strict semantics
> For example, Node.js does not follow "module"
> also there are issues with named exports interop
> if we followed it by default, many packages would break
> because they don't support proper ESM semantics, but what we call "faux ESM" semantics
> that is - bundler magic
> so to build an ecosystem with real es module semantics
> requires a new field
> ie "exports"

**Comment by jneurock** [comment]
I want to make a small change to the Mirage GraphQL code anyway, to remove a dependency on a private-ish export from Mirage. I'll also try removing the `b2a` dependency in the process and publish a new version. See if that works in the REPL.

**Comment by samselikoff** [comment]
Think we can close this now eh?

**Comment by jneurock** [comment]
Yeah? Would be great. On to the next thing!

**Comment by samselikoff** [comment]
I guess we can top it off with https://github.com/miragejs/site/issues/962 and then :chefs-kiss: update the guide with the most super simple example, then close this beast!

**Comment by samselikoff** [comment]
Amazing work on all this Rocky.

**Comment by jneurock** [comment]
Thanks, Sam! Sure, #962 sounds good for this topic. What part of the guide would you update when that's done?

**Comment by samselikoff** [comment]
Just feature it somehow as a link in the graphql page https://miragejs.com/docs/advanced/graphql/#gatsby-focus-wrapper

I think people will be very impressed when they can see it in action!

Also, would be great one day to be able to "embed" repls.

**Comment by jneurock** [comment]
> Just feature it somehow as a link in the graphql page https://miragejs.com/docs/advanced/graphql/#gatsby-focus-wrapper
> 
> I think people will be very impressed when they can see it in action!
> 
> Also, would be great one day to be able to "embed" repls.

Yeah, that makes sense, dropping a link into the page in the guides.

+1 on embedding REPLs in the docs.
